docs/getting-started/throwing-errors.md

```diff
@@ -0,0 +1,289 @@
+
+Some operational errors may be recoverable (as in the page can continue to render even if they occur). Some operational errors may not be recoverable, and in the context of a web service these will result in a non-`200` status code for an end-user. Both of these cases are still "Operational" if they're part of the expected running of the app.
+
+Operational errors should never result in an application crashing unexpectedly.
+
+### Programmer errors
+
+An error is considered to be a "Programmer Error" (or "Non-Operational") if it is unexpected and occurs because of a bug in the code. This category of error can always be avoided or fixed by making a code change. These errors can not be handled properly since the code in question is broken, e.g. `tried to call property X of "undefined"` – this is a result of code that doesn't properly validate the data it handles.
+
+
+## Types of error in code
+
+Reliability Kit provides a package to help us differentiate between Operational (known) errors, and Programmer (unknown) errors: [`@dotcom-reliability-kit/errors`](https://github.com/Financial-Times/dotcom-reliability-kit/tree/main/packages/errors#readme).
+
+The `OperationalError` class (explained fully in the [package documentation](https://github.com/Financial-Times/dotcom-reliability-kit/tree/main/packages/errors#operationalerror)) allows you to throw errors which are explicitly marked as "known". Any errors created with this constructor will include an `isOperational` property set to `true`, which can be used by our tooling later to help us understand the health of a system.
+
+```js
+throw new OperationalError({
+    code: 'EXAMPLE_ERROR',
+    message: 'This is an operational error'
+});
+```
+
+This (and further error classes that which extend it) should ideally be used _anywhere_ where you understand how to handle an error. This means in your app you might do something like this:
+
+```js
+// E.g. GET https://your-app/pokemon/oddish
+app.get('/pokemon/:name', async (request, response, next) => {
+    try {
+        // We make a request to an API, we understand that sometimes this API will fail
+        const response = await fetch(
+            `https://pokeapi.co/api/v2/pokemon-species/${request.params.name}`
+        );
+
+        // If the API request failed, we understand that this is an issue with a remote
+        // service which is expected occasionally as part of running our app. Let's throw
+        if (!response.ok) {
+            throw new OperationalError({
+                code: 'POKEMON_API_FAILED',
+                message: `The Pokemon API responded with a ${response.status} status`,
+                relatesToSystems: ['pokeapi']
+            });
+        }
+
+        response.send(await response.json());
+
+    } catch (error) {
+        // Pass any errors onto the Express error handler
+        next(error);
```

Comment:
**suggestion:** also, it would be clean to have a generic [error handler](https://github.com/Financial-Times/next-newsletters-page/blob/abd26fbb5233b75b1b4f56b76c5adf1f336d7666/server/controllers/displayPage.js#L97) for [operational errors](https://github.com/Financial-Times/next-newsletters-page/blob/abd26fbb5233b75b1b4f56b76c5adf1f336d7666/server/controllers/errorHandler.js#L5)

Reply:
I think this is covered by the [bubbling up in express](https://github.com/Financial-Times/dotcom-reliability-kit/blob/getting-started-guide/docs/getting-started/handling-errors.md#bubbling-up-in-express) section. It's better to pass the errors onto `next` because they will be caught by an Express error handler later.

I haven't documented it yet because it's in-progress, but [we're working on some built-in error handlers](https://github.com/Financial-Times/dotcom-reliability-kit/discussions/26) which _do_ properly handle errors, so you'll be able to add the following to your app:

```js
const createErrorLoggingMiddleware = require('@dotcom-reliability-kit/middleware-log-errors');

app.use(createErrorLoggingMiddleware());
```

In order to take advantage of these we'd have to be using the `next` function in a route rather than something custom because otherwise the errors won't make it to central error handlers and will be output by each route.

The example you give still relies on each route importing an error handler function when it could be handing over to Express instead and avoiding having to remember to pass on all the details like `UUID` or `response`.

Reply:
middleware-log-errors npm approach makes sense. in that case, I would resolve this conversation.